src/prefect/serializers.py

```diff
@@ -15,9 +15,245 @@
     from prefect.packaging.base import PackageManifest
     from prefect.results import _Result
 
+import abc
+import base64
+import json
+import warnings
+from typing import Any, Generic, TypeVar
+
+import cloudpickle
+import pydantic
+from pydantic import BaseModel
+from pydantic.json import pydantic_encoder
+from typing_extensions import Literal
+
+from prefect.orion.serializers import register_serializer
+from prefect.utilities.importtools import from_qualified_name, to_qualified_name
+from prefect.utilities.pydantic import add_type_dispatch
+
+D = TypeVar("D")
+
+
+def prefect_json_object_encoder(obj: Any) -> Any:
+    """
+    `JSONEncoder.default` for encoding objects into JSON with extended type support.
+
+    Raises a `TypeError` to fallback on other encoders on failure.
+    """
+    return {
+        "__class__": to_qualified_name(obj.__class__),
+        "data": pydantic_encoder(obj),
+    }
+
+
+def prefect_json_object_decoder(result: dict):
+    """
+    `JSONDecoder.object_hook` for decoding objects from JSON when previously encoded
+    with `prefect_json_object_encoder`
+    """
+    if "__class__" in result:
+        return pydantic.parse_obj_as(
+            from_qualified_name(result["__class__"]), result["data"]
+        )
+    return result
+
+
+@add_type_dispatch
+class Serializer(BaseModel, Generic[D], abc.ABC):
+    """
+    A serializer that can encode objects of type 'D' into bytes.
+    """
+
+    type: str
+
+    @abc.abstractmethod
+    def dumps(self, obj: D) -> bytes:
+        """Encode the object into a blob of bytes."""
+
+    @abc.abstractmethod
+    def loads(self, blob: bytes) -> D:
+        """Decode the blob of bytes into an object."""
+
+    class Config:
+        extra = "forbid"
+
+
+class PickleSerializer(Serializer):
+    """
+    Serializes objects using the pickle protocol.
+
+    If using cloudpickle, you may specify a list of 'pickle_modules'. These modules will
+    be serialized by value instead of by reference, which means they do not have to be
+    installed in the runtime location. This is especially useful for serializing objects
+    that rely on local packages.
+
+    Wraps pickles in base64 for safe transmission.
+    """
+
+    type: Literal["pickle"] = "pickle"
+
+    picklelib: str = "cloudpickle"
+    picklelib_version: str = None
+    pickle_modules: List[str] = pydantic.Field(default_factory=list)
+
```

Comment:
Does this need load_kwargs and dumps_kwargs similar to JSONSerializer?

Reply:
I don't think there are as many useful argument as there are for JSON. For example, I can see users wanting the `sort_keys` setting but all the pickle ones are obscure with the exception of `protocol` which I'd rather have support for directly. I'd prefer to wait and see if there are requests. I was on the fence about including it in JSON even.